server/svix-server/src/core/idempotency.rs

```diff
@@ -0,0 +1,688 @@
+                // returning whether the value was set
+                let set = if let Ok(set) = redis
+                    .set_if_not_exists(&key, &SerializedResponse::Start, expiry_starting())
+                    .await
+                {
+                    set
+                } else {
+                    return Ok(StatusCode::INTERNAL_SERVER_ERROR.into_response());
+                };
+
+                // If the value was not set, first check the cache for a `Finished` cache value. If
+                // it is instead `None` or the value is a `Start` lock, then enter a loop checking
+                // it every 200ms.
+                //
+                // If the loop times out, then reset the lock and proceed to resolve the service.
+                //
+                // If at any point the cache returns an `Err`, then return 500 response
+                if !set {
+                    match redis.get::<SerializedResponse>(&key).await {
+                        Ok(Some(SerializedResponse::Finished {
+                            code,
+                            headers,
+                            body,
+                        })) => {
+                            return Ok(finished_serialized_response_to_reponse(code, headers, body)
+                                .unwrap_or_else(|_| {
+                                    StatusCode::INTERNAL_SERVER_ERROR.into_response()
+                                }))
+                        }
+
+                        Ok(Some(SerializedResponse::Start)) | Ok(None) => {
+                            if let Ok(Some(SerializedResponse::Finished {
+                                code,
+                                headers,
+                                body,
+                            })) = lock_loop(&redis, &key).await
+                            {
+                                return Ok(finished_serialized_response_to_reponse(
+                                    code, headers, body,
+                                )
+                                .unwrap_or_else(|_| {
+                                    StatusCode::INTERNAL_SERVER_ERROR.into_response()
+                                }));
+                            } else {
+                                // Set the lock if it returns `Ok(None)` and continue to resolve
+                                // as normal, but return 500 if the lock cannot be set
+                                if !matches!(
+                                    redis
+                                        .set_if_not_exists(
+                                            &key,
+                                            &SerializedResponse::Start,
+                                            expiry_starting(),
+                                        )
+                                        .await,
+                                    Ok(true)
```

Comment:
This is not actually used and confused me a bit, but I guess it's better than not having it to indicate this path continues.